plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,666 @@
+	m_dataLock.lock();
+
+	m_scaleRatio = newRatio;
+	m_stepSize = (float)s_windowSize / s_overSampling;
+	m_numWindows = (float)m_originalBuffer.size() / m_stepSize - s_overSampling - 1;
+	m_outStepSize = m_scaleRatio * (float)m_stepSize; // float, else inaccurate
+	m_freqPerBin = m_originalSampleRate / s_windowSize;
+	m_expectedPhaseIn = 2. * F_PI * (float)m_stepSize / (float)s_windowSize;
+	m_expectedPhaseOut = 2. * F_PI * (float)m_outStepSize / (float)s_windowSize;
+
+	m_processedBuffer.resize(m_scaleRatio * m_originalBuffer.size(), 0);
+
+	// very slow :(
+	std::fill(m_processedWindows.begin(), m_processedWindows.end(), false);
+	std::fill(m_processedBuffer.begin(), m_processedBuffer.end(), 0);
+
+	m_dataLock.unlock();
+}
+
+// time shifts one window from originalBuffer and writes to m_processedBuffer
+// resources:
+// http://blogs.zynaptiq.com/bernsee/pitch-shifting-using-the-ft/
+// https://sethares.engr.wisc.edu/vocoders/phasevocoder.html
+// https://dsp.stackexchange.com/questions/40101/audio-time-stretching-without-pitch-shifting/40367#40367
+// https://www.guitarpitchshifter.com/
+// https://en.wikipedia.org/wiki/Window_function
+void PhaseVocoder::generateWindow(int windowNum, bool useCache)
+{
+	// declare vars
+	float real, imag, phase, magnitude, freq, deltaPhase = 0;
+	int windowStart = (float)windowNum * m_stepSize;
+	int windowIndex = (float)windowNum * s_windowSize;
+
+	if (!useCache)
+	{ // normal stuff
+		memcpy(m_FFTInput.data(), m_originalBuffer.data() + windowStart, s_windowSize * sizeof(float));
+
+		// FFT
+		fftwf_execute(m_fftPlan);
+
+		// analysis step
+		for (int j = 0; j < s_windowSize / 2; j++) // only process nyquistic frequency
+		{
+			real = m_FFTSpectrum[j][0];
+			imag = m_FFTSpectrum[j][1];
+
+			magnitude = 2. * sqrt(real * real + imag * imag);
+			phase = atan2(imag, real);
+
+			// calculate difference in phase with prev window
+			freq = phase;
+			freq = phase
+				- m_lastPhase[std::max(0, windowIndex + j - s_windowSize)]; // subtract prev pahse to get phase diference
+			m_lastPhase[windowIndex + j] = phase;
+
+			freq -= (float)j * m_expectedPhaseIn; // subtract expected phase
+			// at this point, freq is the difference in phase
+			// between the last phase, having removed the expected phase at this point in the sample
+
+			// this puts freq in 0-2pi. Since the phase difference is proportional to the deviation in bin frequency,
+			// with this we can better estimate the true frequency
+			freq = fmod(freq + F_PI, -2.0f * F_PI) + F_PI;
+
+			// convert phase difference into bin freq mulitplier
+			freq = (float)s_overSampling * freq / (2. * F_PI);
+
+			// add to the expected freq the change in freq calculated from the phase diff
+			freq = (float)j * m_freqPerBin + freq * m_freqPerBin;
+
+			m_allMagnitudes[j] = magnitude;
+			m_allFrequencies[j] = freq;
+		}
+		// write cache
+		memcpy(m_freqCache.data() + windowIndex, m_allFrequencies.data(), s_windowSize * sizeof(float));
```

Comment:
We should be using ``std::copy`` or ``std::copy_n`` instead. It's from the STL and can lead to simpler code. You used ``memcpy`` everywhere else, so you should try and change those too.

Reply:
Done